Learner Detail page & Learner Engagement Timeline

dan-f opened the pull request:
This PR implements the Learner Detail page, which currently only consists of the Learner Engagement Timeline.

* Clicking on a user in the table now takes you to their detail page
* The detail page renders a timeline trends chart of their engagement

@dsjen please review

dsjen commented on analytics_dashboard/static/js/views/trends-view.js:
```diff
@@ -2,8 +2,43 @@ define(['moment', 'nvd3', 'underscore', 'views/chart-view'],
+             *      }
+             */
+            initialize: function (options) {
+                ChartView.prototype.initialize.call(this, options);
```
Can we remove this method?  It seems like we could to the same effect.

dan-f replied:
The method doesn't serve a functional purpose. I added it here to have something to attach the jsdoc comment to. However, on second look, could I have added this to the `initialize` method in `chart-view.js`.

dsjen commented on analytics_dashboard/static/js/views/chart-view.js:
```diff
@@ -30,6 +30,21 @@ define(['d3', 'jquery', 'nvd3', 'underscore', 'utils/utils', 'views/attribute-li
              * rendering.  The map consists of the trend data as 'values' and
              * the trend title as 'key'.
+             *
+             * Part of the contract is that
```
Thank you for this awesome comment!

dan-f replied:
np! :)

dsjen commented on analytics_dashboard/static/js/views/attribute-listener-view.js:
```diff
@@ -8,19 +8,40 @@ define(['backbone'],
                 }
             },
 
+            isDataAvailable: function () {
```
Nit: it almost seems like the models should be required to implement some sort of data status rather than having the view determine if that is available.

dan-f replied:
You're totally right. Lemme take a look at this...

dsjen commented on analytics_dashboard/static/js/views/attribute-listener-view.js:
```diff
@@ -8,19 +8,40 @@ define(['backbone'],
 
+            isDataAvailable: function () {
+                var isDataAvailableFunc = this.options.isDataAvailable;
+                if (typeof isDataAvailableFunc === 'function') {
```
Nit: We could use underscore instead, http://underscorejs.org/#isFunction.

dan-f replied:
Fair enough. I didn't want to introduce underscore for such a small use case, but since Backbone already requires it, why not!

dsjen commented on analytics_dashboard/static/apps/learners/js/views/engagement-timeline.js:
```diff
@@ -0,0 +1,52 @@
+    'marionette',
+    'text!learners/templates/engagement-timeline.underscore',
+    'views/trends-view',
+    'underscore'
```
I guess `underscore`' should come before `views` to be alphabetically compliant. :)

dan-f replied:
correct!

dsjen commented on analytics_dashboard/static/apps/learners/js/utils.js:
```diff
@@ -47,6 +47,71 @@ define([], function () {
+                if (status === 504) {
+                    return [
+                        'appError',
+                        gettext('504: Server error: processing your request took too long to complete. Reload the page to try again.') // jshint ignore:line
```
Nit: A translator comment about 504 might be helpful.

dan-f replied:
Got it.

dsjen commented on analytics_dashboard/static/apps/learners/js/spec/controller-spec.js:
```diff
@@ -3,19 +3,22 @@ define([
+        var courseId,
+            server;
+
+        courseId = 'test/course/id';
```
Nit: This could be moved up to line 11.

dan-f replied:
If you don't mind, I think I'll leave it. Lately I've found code easier to refactor when the variable declarations are decoupled from their assignments.

dsjen replied:
:+1:

dsjen commented:
It seems like hiding/showing trends using the legend changes the styles.  I'd like to touch base on that post weekend.  Otherwise, this looks great!

dan-f commented:
@dsjen I addressed your feedback in my latest commit. Please let me know if there's anything else you need me to do. Let's figure out the styling issue on Monday.

dsjen commented:
:dancer: :dancers: :+1: :100: :cake: :v:
